server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java

```diff
@@ -220,15 +234,54 @@ public CompletionStage<Collection<CacheDistributionInfo>> cacheDistribution(Stri
          return CompletableFuture.completedStage(Collections.singletonList(local));
       }
 
-      Map<Address, CacheDistributionInfo> distributions = new ConcurrentHashMap<>();
+      return requestMembers(dm.getCacheTopology().getMembers(), ecm -> {
+         Cache<?, ?> c = SecurityActions.getCache(cacheName, ecm);
+         return CacheDistributionInfo.resolve(c.getAdvancedCache());
+      });
+   }
+
+   public CompletionStage<CompleteKeyDistribution> getKeyDistribution(String cacheName, Object key, RestRequest request) {
+      AdvancedCache<Object, ?> cache = getCache(cacheName, request.keyContentType(), request);
+      checkCachePermission(cache, request.getSubject(), AuthorizationPermission.MONITOR);
+      ByRef.Boolean containsKey = new ByRef.Boolean(false);
+      return SecurityActions.cacheContainsKeyAsync(cache, key)
+            .thenCompose(contains -> {
+               containsKey.set(contains);
+
+               DistributionManager dm = SecurityActions.getDistributionManager(cache);
+               if (dm == null) {
+                  KeyDistributionInfo local = KeyDistributionInfo.resolve(cache, true);
+                  return CompletableFuture.completedFuture(Collections.singletonList(local));
+               }
+
+               LocalizedCacheTopology topology = dm.getCacheTopology();
+               List<Address> members = topology.getMembers();
+               Address primary = null;
+               if (contains) {
+                  DistributionInfo distribution = topology.getDistribution(key);
+                  members = distribution.readOwners();
+                  primary = distribution.primary();
+               }
+
+               final Address p = primary;
+               return requestMembers(members, ecm -> {
```

Comment:
I don't think I will ever like this, but I guess it is alright. I just hope this distribution info is not queried N times for some operation.